packages/block-editor/src/store/selectors.js

```diff
@@ -463,6 +464,48 @@ export function getBlockHierarchyRootClientId( state, clientId ) {
 	return parent;
 }
 
+/**
+ * Given a block client ID, returns the next element of the hierarchy from which the block is nested which should be selected onFocus, return the block itself for root level blocks.
+ *
+ * @param {Object} state    Editor state.
+ * @param {string} clientId Block from which to find first to select client ID.
+ *
+ * @return {string} First to select client ID
+ */
+export function getLowestCommonAncestorWithSelectedBlock( state, clientId ) {
```

Comment:
Let's also add some unit tests for this one.

Also, when I was trying to understand these better the other day, I came up with an alternative that looks simpler and faster. Because this is a tree, every two blocks have at least one common ancestor: the root node. Here we don't really have a concept of the root node and we're representing it as `undefined`, but the approach also works.

Instead of doing a bottom-up lookup, we can start at the top. The first divergence you see on the path to each node is when you passed the lowest common ancestor.

```js
const clientParents = [ ...getBlockParents( state, clientId ), clientId ];
const selectedParents = [ ...getBlockParents( state, selectedId ), selectedId ];
let lca = undefined; // No common ancestor

const maxDepth = Math.min( clientParents.length, selectedParents.length );
for (let index = 0; index < maxDepth; index++) {
	if ( clientParents[index] === selectedParents[index] ) {
		lca = clientParents[index];
	} else {
		break;
	}
}
return lca;
```

Reply:
Nice refactor there, thanks. Also will give it a shot after deal with styles thing

Reply:
I did the refactor and will be adding some unit test

Reply:
I have added unit test for `getLowestCommonAncestorWithSelectedBlock()`selector